### PR TITLE
Create a unitary to pauli string transformer

### PR DESCRIPTION
**Pull request by NoureldinYosri**
- This is the first step to implementing the algorithm in https://github.com/quantumlib/Cirq/issues/5906#issuecomment-1551754815
- Based on https://quantumcomputing.stackexchange.com/questions/13157/how-do-i-check-if-a-gate-represented-by-unitary-u-is-a-clifford-gate#comment17495_13158
- The complexity of this algorithm is $\mathcal{O}(n 4^n)$ where $n$ is the number of qubits it acts on or alternatively $\mathcal{O}(S \log_4{S})$ where $S$ is the size of the unitary.

### CODE REVIEW

File: cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    U = U.T
+    for i in range(U.shape[1]):
+        _fast_walsh_hadamard_transform(U[:, i])
+    return U
+
+
+def unitary_to_pauli_string(U: npt.NDArray, eps: float = 1e-15) -> Optional[DensePauliString]:
+    """Attempts to find a pauli string (with possible phase) equivalent to U up to eps.
+
+        Based on this answer https://shorturl.at/aA079.
```
Comment (dstrain115) [inline cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition.py:64]: I think it might be worth restating the algorithm here as well as linking to the source.  This is pretty hard to follow otherwise.

Maybe add some line comments below too.

File: cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition_test.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    got = unitary_to_pauli_string(np.array([[1, 1], [0, 2]]))
+    assert got is None
+
+    got = unitary_to_pauli_string(np.array([[0, 0.5], [1, -1]]), eps=1.1)
+    assert got is None
+
+
+def test_invalid_input():
+    with pytest.raises(ValueError):
```
Comment (dstrain115) [inline cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition_test.py:54]: nit; consider adding a match here in case a second type of value error is added later.

File: cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition_test.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+@pytest.mark.parametrize('phase', [cmath.exp(i * 2 * cmath.pi / 5 * 1j) for i in range(5)])
+@pytest.mark.parametrize(
+    'pauli_string', [''.join(p) for p in itertools.product(['', 'I', 'X', 'Y', 'Z'], repeat=4)]
+)
+def test_unitary_to_pauli_string(pauli_string: str, phase: complex):
+    want = DensePauliString(pauli_string, coefficient=phase)
+    got = unitary_to_pauli_string(protocols.unitary(want))
+    assert got is not None
+    assert np.all(want.pauli_mask == got.pauli_mask)
+    assert np.isclose(cast(np.complex128, want.coefficient), cast(np.complex128, got.coefficient))
```
Comment (dstrain115) [inline cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition_test.py:36]: Should we add a test for eps != 0 as well?

File: cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+from typing import Optional, Tuple, cast
+
+import numpy as np
+import numpy.typing as npt
+
+from cirq.ops import DensePauliString
+from cirq import protocols
+
+
+def _argmax(V: npt.NDArray) -> Tuple[int, float]:
```
[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition.py:24]: done

File: cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition.py (hunk 1 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition.py:64]: done

File: cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition_test.py (hunk 2 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition_test.py:54]: added

File: cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition_test.py (hunk 3 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition_test.py:36]: the default value is `eps=1e-15` not zero.

**Comment by NoureldinYosri** [comment]
@dstrain115 PTAL

**Comment by NoureldinYosri** [comment]
@dstrain115 ping

**Review by dstrain115 — APPROVED** [review]

File: cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+def unitary_to_pauli_string(U: npt.NDArray, eps: float = 1e-15) -> Optional[DensePauliString]:
+    """Attempts to find a pauli string (with possible phase) equivalent to U up to eps.
+
+        Based on this answer https://shorturl.at/aA079.
+        Let x_mask be the index of the maximum number of the first column of U
+        and z_mask be the index of the maximum number of the first column of H†UH
+        each of these indicies is n-bits long where U is 2^n x 2^n.
+
+        These two indicies/masks encode in binary the indicies of the qubits that
```
[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition.py:71]: done

File: cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+    Returns:
+        A DensePauliString of None.
+
+    Raises:
+        ValueError: if U is not square with a power of 2 dimension.
+    """
+
+    if len(U.shape) != 2 or U.shape[0] != U.shape[1]:
+        raise ValueError(f'Input has a none square shape {U}')
```
[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition.py:90]: done
